tools/rsfont/convert_png.c

```diff
@@ -1,169 +1,131 @@
-		FATAL_ERROR("Failed to init I/O for reading \"%s\".\n", path);
-
-	png_init_io(png_ptr, fp);
-	png_set_sig_bytes(png_ptr, 8);
-	png_read_info(png_ptr, info_ptr);
-
-	int bit_depth = png_get_bit_depth(png_ptr, info_ptr);
-
-	if (bit_depth != image->bitDepth)
-		FATAL_ERROR("\"%s\" has a bit depth of %d, but the expected bit depth is %d.\n", path, bit_depth, image->bitDepth);
-
-	int color_type = png_get_color_type(png_ptr, info_ptr);
-
-	if (color_type != PNG_COLOR_TYPE_GRAY && color_type != PNG_COLOR_TYPE_PALETTE)
-		FATAL_ERROR("\"%s\" has an unsupported color type.\n", path);
+    if (bit_depth != image->bitDepth)
+        FATAL_ERROR("\"%s\" has a bit depth of %d, but the expected bit depth is %d.\n", path, bit_depth, image->bitDepth);
+    if (color_type != LCT_GREY && color_type != LCT_PALETTE)
+        FATAL_ERROR("\"%s\" has an unsupported color type.\n", path);
 
 	// Check if the image has a palette so that we can tell if the colors need to be inverted later.
 	// Don't read the palette because it's not needed for now.
-	image->hasPalette = (color_type == PNG_COLOR_TYPE_PALETTE);
+    image->hasPalette = (color_type == LCT_PALETTE);
+
+    image->width = width;
+    image->height = height;
 
-	image->width = png_get_image_width(png_ptr, info_ptr);
-	image->height = png_get_image_height(png_ptr, info_ptr);
+    free(buffer);
+}
 
-	int rowbytes = png_get_rowbytes(png_ptr, info_ptr);
+void ReadPngPalette(char *path, struct Palette *palette)
+{
+    // LodePNG doesn't seem to provide a method to read the palette without
+    // decoding the whole file.
+    LodePNGState state;
+    unsigned int error;
+    int filesize;
+    unsigned char *buffer = ReadWholeFile(path, &filesize);
+    unsigned char *dummy;
+    unsigned int dummy_w, dummy_h;
 
-	image->pixels = malloc(image->height * rowbytes);
+    lodepng_state_init(&state);
 
-	if (image->pixels == NULL)
-		FATAL_ERROR("Failed to allocate pixel buffer.\n");
+    state.decoder.color_convert = 0;
 
-	png_bytepp row_pointers = malloc(image->height * sizeof(png_bytep));
+    error = lodepng_decode(&dummy, &dummy_w, &dummy_h, &state, buffer, filesize);
+    if (error)
+        FATAL_ERROR("Failed to decode image \"%s\": %s\n", path, lodepng_error_text(error));
 
-	if (row_pointers == NULL)
-		FATAL_ERROR("Failed to allocate row pointers.\n");
+    if (state.info_png.color.colortype != LCT_PALETTE)
+        FATAL_ERROR("The image \"%s\" does not contain a palette.\n", path);
 
-	for (int i = 0; i < image->height; i++)
-		row_pointers[i] = (png_bytep)(image->pixels + (i * rowbytes));
+    if (state.info_png.color.palettesize > 256)
+        FATAL_ERROR("Images with more than 256 colors are not supported.\n");
 
-	if (setjmp(png_jmpbuf(png_ptr)))
-		FATAL_ERROR("Error reading from \"%s\".\n", path);
+    palette->numColors = state.info_png.color.palettesize;
 
-	png_read_image(png_ptr, row_pointers);
+    // TODO: Why is LodePNG getting the palette size wrong?
```

Comment:
I don't know if it is getting it wrong. There appear to be some PNG files with more colors in their palette than they should have. For instance, "graphics/trade/gba.png" has 64 colors in its PLTE chunk even though it should have 16.